rfcs/text/0006_management_section_service.md

```diff
@@ -0,0 +1,343 @@
+}
+```
+
+We can also create a utility in `kibana_react` to make it easy for folks to `mount` a React app:
+```ts
+// src/plugins/kibana_react/public/mount_with_react.tsx
+import { KibanaContextProvider } from './context';
+
+export const mountWithReact = (
+  Component: React.ComponentType<{ basename: string }>,
+  context: AppMountContext,
+  params: ManagementSectionMountParams,
+) => {
+  ReactDOM.render(
+    (
+      <KibanaContextProvider services={{ ...context }}>
+        <Component basename={params.sectionBasePath} />
+      </KibanaContextProvider>
+    ),
+    params.element
+  );
+
+  return () => ReactDOM.unmountComponentAtNode(params.element);
+}
+
+// my_plugin/public/plugin.ts
+import { mountWithReact } from 'src/plugins/kibana_react/public';
+
+export class MyPlugin {
+  setup(core, { management }) {
+    const kibanaSection = management.sections.get('kibana');
+    kibanaSection.registerLink({
+      id: 'my-other-kibana-section-link',
+      ...,
+      async mount(context, params) {
+        const { MySection } = await import('./components/my-section');
+        return mountWithReact(MySection, context, params);
+      }
+    });
+  }
+}
+```
+ 
+# Detailed design
+
+```ts
+interface ManagementSetup {
+  sections: SectionsServiceSetup;
+}
+
+interface ManagementStart {
+  sections: SectionsServiceStart;
+}
+ 
+interface SectionsServiceSetup {
+  get: (sectionId: string) => Section;
+  getAvailable: () => Section[]; // filtered based on capabilities
+  register: Register;
+}
+
+interface SectionsServiceStart {
+  getAvailable: () => Array<Omit<Section, 'registerLink'>>; // filtered based on capabilities
+  // uses `core.application.navigateToApp` under the hood, automatically prepending the `path` for the link
+  navigateToSectionLink: (linkId: string, options?: { path?: string; state?: any }) => void;
+}
+
+type Register = (
+  id: string,
+  title: string,
+  description: string, // not used in current UI, but possibly in future
+  order?: number,
+  euiIconType?: string, // takes precedence over `icon` property.
+  icon?: string, // URL to image file; fallback if no `euiIconType`
+) => Section | Error;
+ 
+type RegisterLink = (
+  id: string;
+  title: string;
+  description: string; // not used in current UI, but possibly in future
+  order?: number;
+  mount: ManagementSectionMount;
+) => Link | Error;
+ 
+type Unmount = () => Promise<void> | void;
+ 
+interface ManagementSectionMountParams {
+  sectionBasePath: string; // base path for setting up your router
+  element: HTMLElement; // element the section should render into
+}
+
+type ManagementSectionMount = (
+  context: AppMountContext, // provided by core.ApplicationService
+  params: ManagementSectionMountParams,
+) => Unmount | Promise<Unmount>;
+
+interface Link {
+  id: string;
+  title: string;
+  description: string;
+  basePath: string;
+  sectionId: string;
+  order?: number;
+  destroy: () => Promise<void> | void; // de-registers & calls unmount()
```

Comment:
What's the usecase for these destroy methods? If we didn't want to show a section or link couldn't we just not register it?

Reply:
Good question; this was something I wasn't sure about. I kept it here because the legacy API has a concept of "deregistration," but after looking at it again, it seems it was mainly used as a workaround for the lack of `capabilities`.

I can't really think of a good use case where another application would want/need to hide a particular section outside of `capabilities`, so I'll remove it for now.